src/ripple/app/misc/detail/WorkBase.h

```diff
@@ -0,0 +1,221 @@
+    std::string port_;
+    callback_type cb_;
+    boost::asio::io_service& ios_;
+    boost::asio::io_service::strand strand_;
+    resolver_type resolver_;
+    socket_type socket_;
+    request_type req_;
+    response_type res_;
+    beast::streambuf read_buf_;
+
+public:
+    WorkBase(
+        std::string const& host, std::string const& path,
+        std::string const& port,
+        boost::asio::io_service& ios, callback_type cb);
+    ~WorkBase();
+
+    Impl&
+    impl()
+    {
+        return *static_cast<Impl*>(this);
+    }
+
+    void run() override;
+
+    void cancel() override;
+
+    void
+    fail(error_code const& ec);
+
+    void
+    on_resolve(error_code const& ec, resolver_type::iterator it);
+
+    void
+    on_start();
+
+    void
+    on_request(error_code const& ec);
+
+    void
+    on_response(error_code const& ec);
+};
+
+//------------------------------------------------------------------------------
+
+template<class Impl>
+WorkBase<Impl>::WorkBase(std::string const& host,
+    std::string const& path, std::string const& port,
+    boost::asio::io_service& ios, callback_type cb)
+    : host_(host)
+    , path_(path)
+    , port_(port)
+    , cb_(std::move(cb))
+    , ios_(ios)
+    , strand_(ios)
+    , resolver_(ios)
+    , socket_(ios)
+{
+}
+
+template<class Impl>
+WorkBase<Impl>::~WorkBase()
+{
+    if (cb_)
+        cb_ (make_error_code(boost::system::errc::not_a_socket),
+            std::move(res_));
+}
+
+template<class Impl>
+void
+WorkBase<Impl>::run()
+{
+    if (! strand_.running_in_this_thread())
+        return ios_.post(strand_.wrap (std::bind(
+            &WorkBase::run, impl().shared_from_this())));
+
+    resolver_.async_resolve(
+        query_type{host_, port_},
+        strand_.wrap (std::bind(&WorkBase::on_resolve, impl().shared_from_this(),
+            beast::asio::placeholders::error,
+                beast::asio::placeholders::iterator)));
+}
+
+template<class Impl>
+void
+WorkBase<Impl>::cancel()
+{
+    if (! strand_.running_in_this_thread())
+    {
+        return ios_.post(strand_.wrap (std::bind(
+            &WorkBase::cancel, impl().shared_from_this())));
+    }
+
+    error_code ec;
+    resolver_.cancel();
+    socket_.cancel (ec);
+}
+
+template<class Impl>
+void
+WorkBase<Impl>::fail(error_code const& ec)
+{
+    if (cb_)
+    {
+        cb_(ec, std::move(res_));
+        cb_ = nullptr;
+    }
+}
+
+template<class Impl>
+void
+WorkBase<Impl>::on_resolve(error_code const& ec, resolver_type::iterator it)
+{
+    if (ec)
+        return fail(ec);
+
+    socket_.async_connect(*it,
+        strand_.wrap (std::bind(&Impl::on_connect, impl().shared_from_this(),
+            beast::asio::placeholders::error)));
+}
+
+template<class Impl>
+void
+WorkBase<Impl>::on_start()
+{
+    req_.method = "GET";
+    req_.url = path_;
+    req_.version = 11;
+    req_.headers.replace (
+        "Host", host_ + ":" + port_);
+    req_.headers.replace ("User-Agent", "Beast");
```

Comment:
Should probably be `rippled` and a version. For example: `req_.headers.replace ("User-Agent", BuildInfo::getFullVersionString());` will give a user agent like: *rippled-0.33.0-b6*.